### PR TITLE
[OP] Move reshape and flatten to nnvm

### PR DESCRIPTION
**Pull request by sxjscience**
The FCompute directly calls IdentityCompute while the backward node uses _backward_copy. Also, **nnvm::Tuple\<int\>** is used as the type of shape, which may contain negative values.

### CODE REVIEW

**Comment by piiswrong** [comment]
Do we have tests for flatten and reshape?

File: src/operator/tensor/matrix_op-inl.h [hunk 1]
```diff
@@ -14,6 +14,189 @@
 namespace mxnet {
 namespace op {
 
+struct ReshapeParam : public dmlc::Parameter<ReshapeParam> {
+  TShape target_shape;
+  bool keep_highest;
+  nnvm::Tuple<int> shape;
```
Comment (tqchen) [inline src/operator/tensor/matrix_op-inl.h]: nnvm tuple is fine, it is needed in shape inference

**Comment by sxjscience** [comment]
We have test for Reshape but not for Flatten.

**Comment by sxjscience** [comment]
@piiswrong Test of Flatten has been added

**Pull request merged**